### PR TITLE
Add before_validation hooks to resources

### PR DESCRIPTION
**Pull request by A-Boudi**
Solves #122 

This adds a before_validation hook to be fired just before the before_commit hook.

### CODE REVIEW

File: spec/integration/rails/transaction_hooks_spec.rb [hunk 1]
```diff
@@ -230,13 +238,31 @@ def json
...
         end
       end
+
+      context "when an error is raised before_validation" do
+        before do
+          $raise_on_before_validation = {employee: true}
+        end
+
+        it "does not run before_commit callbacks" do
+          expect_any_instance_of(Graphiti::Util::ValidationResponse)
```
Comment (richmolj) [inline spec/integration/rails/transaction_hooks_spec.rb:256]: You did a great job with the general setup, but I think this is the issue that would hold up the PR. We wanted something to fire ***before*** validation. As you can see and line 126 of `resource_proxy.rb`, this is happening ***after*** validation. IOW, we want `to_not receive` - and probably a Rails integration spec that addresses the [original issue](https://github.com/graphiti-api/graphiti/issues/93).

I don't think you're too far off though! Hope we get this in 👏 !

**Comment by A-Boudi** [comment]
Hi, @richmolj, I'm not sure if I did understand the issue correctly.

Here is what I originally understood:
During the persistance of a resource, graphiti works like this:
* Process resource's parents (mainly belongs_to relationships)
* Persist the resource with the foreing keys retrieved above
* Process resource's children (has_many and many_to_many relationships). Persist the children using the resource key.

If a resource has validation on its children, it can't be saved because the children are not accessible when the resource is being saved.

To fix this issue, the validation on children are scoped via a `:whatever` context. And a second round of validation is done using `:whatever` after that all the resource's parents and children are persisted and accessible.

Now, if we want something to fire before **validation**. Would this mean that this new callback should be fired before `before_save` callback?

**Comment by A-Boudi** [comment]
Hi @richmolj, what do you think?

**Comment by richmolj** [comment]
@A-Boudi thanks for the reminder and sorry for the delay. I think you have correct understanding of the issue, but the word `validation` is tripping things up. Graphiti "validation" is actually just checking model, instances for `#errors` - we expect the actual validation to be done with ActiveModel and performed during save.

So, this is maybe this is better named as `before_validation_processing` or `after_graph_persisted` or something like that.

Still, it should fire before `Graphiti::Util::ValidationResponse` code fires. Otherwise we'd be performing the validation, but nothing would happen in response (no rolling back the transaction or rendering errors correctly).

**Comment by A-Boudi** [comment]
Hi @richmolj,

Like asked, I moved this hook to fire before graphiti validation. I also changed the name to `after_graph_persist`.

For the tests, I have kept the test introduced in the first commit. And I added new integration tests that addresses the [original issue](#93 ).

**Comment by richmolj** [comment]
Really, ***REALLY*** great job @A-Boudi 💯 🥇 ! Thanks so much for sticking with it - I really appreciate the integration tests in particular ❤️  If you could please add an entry to the CHANGELOG, I'd be happy to merge!

**Comment by A-Boudi** [comment]
@richmolj CHANGELOG updated!

**Pull request merged**